File: medium/210. Course Schedule II/210. Course Schedule II.ts
function findOrder(numCourses: number, prerequisites: number[][]): number[] {
    const graph: { [key in number]: number[] } = Object.fromEntries(
        Array.from({ length: numCourses }, (_, i) => [i, []])
    )

    for (const [course, pre] of prerequisites) {
        graph[course].push(pre)
    }

    const state = new Array(numCourses).fill(0)
    const visited = new Set()

    const hasCycle = (course: number): [boolean, number] => {
        if (!graph[course].length)
            return [true, state[course]]
        if (visited.has(course))
            return [false, Infinity]

        visited.add(course)
        for (const pre of graph[course]) {
            const [doable, order] = hasCycle(pre)
            if (!doable) return [false, Infinity]
            state[course] = Math.max(state[course], 1 + order)
        }

        graph[course] = []
        return [true, state[course]]
    }

    for (let course = 0; course < numCourses; course++) {
        const [doable, _] = hasCycle(course)
        if (!doable)
            return []
    }

    return Array.from({ length: numCourses }, (_, i) => state[i]) // .sort((a, b) => state[a] - state[b])
};

export { }

File: medium/210. Course Schedule II/210. Course Schedule II.py
from typing import List, Tuple


class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = {i: [] for i in range(numCourses)}
        for course, pre in prerequisites:
            graph[course].append(pre)

        state = [0] * numCourses
        course_order = set()

        def has_cycle(course: int) -> Tuple[bool, int]:
            if not graph[course]:
                return True, state[course]
            if course in course_order:
                return False, float("Infinity")

            course_order.add(course)
            for pre in graph[course]:
                doable, order = has_cycle(pre)
                if not doable:
                    return False, float("Infinity")
                state[course] = max(state[course], 1 + order)

            graph[course] = []

            return True, state[course]

        for course in range(numCourses):
            doable, _ = has_cycle(course)
            if not doable:
                return []

        return sorted(list(range(numCourses)), key=lambda i: state[i])


solution = Solution()
print(solution.findOrder(numCourses=2, prerequisites=[[1, 0]]))
print(solution.findOrder(numCourses=4, prerequisites=[[1, 0], [2, 0], [3, 1], [3, 2]]))
print(solution.findOrder(numCourses=4, prerequisites=[[0, 1], [0, 2], [3, 1], [3, 2]]))
print(solution.findOrder(numCourses=1, prerequisites=[]))
